STORM-2018: Supervisor V2.

revans2 opened the pull request:
Still needs run as user and CGroup work, but the rest is working

Any feedback on this would be welcome.  I am particularly interested in the readability of the code and how easy is it to understand vs the original supervisor.

All the unit tests pass and I have done a lot of manual testing switching back and forth between the original supervisor and supervisor V2 to ensure that this can be a rolling upgrade.

There are still a number of TODOs in the code.  Most of them are for removing the original supervisor code and cleaning up the result of that.

knusbaum commented on storm-core/src/jvm/org/apache/storm/daemon/supervisor/Container.java:
```diff
@@ -0,0 +1,417 @@
+     * @throws IOException
+     */
+    protected void forceKill(long pid) throws IOException {
+        Utils.forceKillProcess(String.valueOf(pid));
```
Do we want to log a force kill like we do a "nice" kill?

knusbaum commented on storm-core/src/jvm/org/apache/storm/daemon/supervisor/Container.java:
```diff
@@ -0,0 +1,417 @@
+                if (user.equals(processUser)) {
+                    return true;
+                } else {
+                    LOG.info("Found {} running as {}, but expected it to be {}", pid, processUser, user);
```
break; ?

knusbaum commented on storm-core/src/jvm/org/apache/storm/daemon/supervisor/Container.java:
```diff
@@ -0,0 +1,417 @@
+        if (Utils.checkFileExists(workerDir)) {
+            LOG.debug("Creating symlinks for worker-id: {} topology-id: {} to its port artifacts directory", _workerId, topologyId);
+            Utils.createSymlink(workerDir, topoDir, "artifacts", String.valueOf(_port));
+        }
```
Failure here seems like something we might want to at least log.

revans2 replied:
I agree the code way copied and pasted from before, but we should have eliminated the races so if there worker dir is not there we have a bug and should do something about it.

d2r replied:
In `writeLogMetadata` we actually try to set up the directory if we don't find it first. Then we write the log metadata there. Is this a similar case when if we find that the dir is not there already we should throw?

knusbaum commented:
+1 on general direction. Readability is good; classes aren't too large or convoluted.

knusbaum commented on storm-core/src/jvm/org/apache/storm/daemon/supervisor/Supervisor.java:
```diff
@@ -17,135 +17,541 @@
-import java.util.concurrent.Callable;
+public class Supervisor implements DaemonCommon, AutoCloseable {
+    
+    public class ReadStateThread implements Runnable, AutoCloseable {
```
This is big enough that I wonder if it should be its own class. I don't see that it's taking advantage of the special access inner classes have to any extent, and we already have a supervisor package it could comfortably live in.

revans2 replied:
That sounds reasonable.

harshach commented on storm-core/src/jvm/org/apache/storm/daemon/supervisor/BasicContainer.java:
```diff
@@ -0,0 +1,494 @@
+    protected final String _profileCmd;
+    protected volatile boolean _exitedEarly = false;
+    
+    private class ProcessExitCallback implements Utils.ExitCodeCallable<Void> {
```
can we move the ExitCodeCallable out of Utils , seems that we can have Processor package .

revans2 replied:
I'll look into moving all of the launchProcess etc to the Supervisor package.  That is the only code that uses it anyways.

harshach commented on storm-core/src/jvm/org/apache/storm/daemon/supervisor/BasicContainer.java:
```diff
@@ -0,0 +1,494 @@
+    }
+
+    private List<String> jmapDumpCmd(String pid, String targetDir) {
+        return Lists.newArrayList(_profileCmd, pid, "jmap", targetDir);
```
should these "jmap" should replaced by ProfileAction.JMAP_DUMP

revans2 replied:
Not sure what you mean by this.  Ideally if we had an enum that we could put a function like cmd in.  Then It would just be

```
return action.cmd(workerPid, targetDir, stop);
```

but ProfileAction is thrift generated code so we cannot play games like that here.

harshach commented on storm-core/src/jvm/org/apache/storm/daemon/supervisor/BasicContainer.java:
```diff
@@ -0,0 +1,494 @@
+                rets.addAll(Arrays.asList(strings));
+            }
+        } else if (value instanceof List) {
+            @SuppressWarnings("unchecked")
```
should we move this to the method itself

revans2 replied:
I put it here because it limits the scope.  That way if someone did something bad in another part of the same method it would still be a warning.

harshach commented on storm-core/src/jvm/org/apache/storm/daemon/supervisor/BasicContainer.java:
```diff
@@ -0,0 +1,494 @@
+        ProcessExitCallback processExitCallback = new ProcessExitCallback(logPrefix);
+        _exitedEarly = false;
+
+        String stormHome = ConfigUtils.concatIfNotNull(System.getProperty("storm.home"));
```
we should make these strings as final string and keep it in Constants file

harshach commented on storm-core/src/jvm/org/apache/storm/daemon/supervisor/BasicContainer.java:
```diff
@@ -0,0 +1,494 @@
+    }
+    
+    @Override
+    public void launch() throws IOException {
```
It will be good add docs around the launch method . In current 1.x branch this code takes sometime to understand

harshach replied:
Also It will be greatly helpful if we can break this into multiple methods.

harshach commented on storm-core/src/jvm/org/apache/storm/daemon/supervisor/Container.java:
```diff
@@ -0,0 +1,429 @@
+     * @throws IOException on any error
+     */
+    protected boolean isProcessAlive(long pid, String user) throws IOException {
+        if (Utils.IS_ON_WINDOWS) {
```
should we break this into two methods one for windows and other for linux

revans2 replied:
done

harshach commented on storm-core/src/jvm/org/apache/storm/daemon/supervisor/Slot.java:
```diff
@@ -0,0 +1,749 @@
+        if (dynamicState.newAssignment != null) {
+            pendingDownload = staticState.localizer.requestDownloadBaseTopologyBlobs(dynamicState.newAssignment.get_topology_id(), staticState.port);
+        }
+        Time.sleep(1000);
```
Any significance to sleep here. we should add a comment and make it config option.

revans2 replied:
You are right that I missed this.

https://github.com/apache/storm/blob/master/storm-core/src/jvm/org/apache/storm/daemon/supervisor/workermanager/DefaultWorkerManager.java#L268

I'll fix it.

harshach commented on storm-core/src/jvm/org/apache/storm/daemon/supervisor/Slot.java:
```diff
@@ -0,0 +1,749 @@
+        assert(dynamicState.container != null);
+        
+        dynamicState.container.kill();
+        Time.sleep(1000);
```
same here

harshach commented on storm-core/src/jvm/org/apache/storm/daemon/supervisor/Slot.java:
```diff
@@ -0,0 +1,749 @@
+            //Scheduling changed after we killed all of the processes
+            return prepareForNewAssignmentOnEmptySlot(cleanupCurrentContainer(dynamicState, staticState, null), staticState);
+        }
+        if ((Time.currentTimeMillis() - dynamicState.startTime) > 120000) {
```
the value here should be driven by config

revans2 replied:
Too many configs make it very confusing.  I would prefer to just leave it hard coded unless someone has a really valid use case where we would want to make it configurable.

harshach replied:
@revans2 agree on too many configs . But can we add atleast some comment on why the value 120000 being used.

harshach commented on storm-core/src/jvm/org/apache/storm/daemon/supervisor/Supervisor.java:
```diff
@@ -17,135 +17,541 @@
+                retries.set(0);
+                return portLA;
+            } catch (RuntimeException e) {
+                if (retries.get() > 2) {
```
this should be config

revans2 replied:
This was hard coded before and I just moved the code over.  I also don't see a lot of value in making it configurable at this point.

harshach commented:
@revans2 gave a first pass.  In terms of readability this looks great. Left minor comments and would like to see supervisor.lanuch() command broken down to smaller pieces if possible and some docs around it.

revans2 commented:
@harshach @knusbaum I think I addressed all of your review comments along with upmerging to the latest master.  I still have a bit more to do, mostly with writing a replacement for supervisor_test.clj but it is fairly close to what I think is a final solution.

revans2 commented on storm-core/src/jvm/org/apache/storm/daemon/supervisor/AdvancedFSOps.java:
```diff
@@ -0,0 +1,202 @@
+            SupervisorUtils.processLauncherAndWait(_conf, user, Arrays.asList("blob", path.toString()), null, logPrefix);
+        }
+        
+        public void deleteIfExists(File path, String user, String logPrefix) throws IOException {
```
need Override tag

revans2 commented on storm-core/src/jvm/org/apache/storm/daemon/supervisor/AdvancedFSOps.java:
```diff
@@ -0,0 +1,202 @@
+            }
+        }
+        
+        public void setupStormCodeDir(Map<String, Object> topologyConf, String path) throws IOException {
```
need Override tag

d2r commented on storm-core/src/jvm/org/apache/storm/cluster/StormClusterStateImpl.java:
```diff
@@ -164,21 +164,18 @@ public Assignment assignmentInfo(String stormId, Runnable callback) {
-        return map;
+        VersionedData<byte[]> dataWithVersion = stateStorage.get_data_with_version(ClusterUtils.assignmentPath(stormId), callback != null);
+        if (dataWithVersion != null) {
+            assignment = ClusterUtils.maybeDeserialize((byte[]) dataWithVersion.getData(), Assignment.class);
```
VersionedData is generic, so we can remove the cast to byte[]

d2r commented on storm-core/src/jvm/org/apache/storm/container/ResourceIsolationInterface.java:
```diff
@@ -56,4 +64,13 @@
+     * other ways of getting PIDs. An Empty set or null if
+     * no PIDs are found.
+     * @throws IOException on any error
+     */
```
Let's just return a Set and have it be empty if no PIDs are found. Callers then don't have to check both cases.

revans2 replied:
There is only one caller currently, and I thought because this is pluggable and the caller is not, that being more lenient here would be better, but I am happy to change it if you feel strongly about it.

srdo replied:
I agree with @d2r, it's just as easy for the implementation to return Collections.emptySet() as null, and null checking can be easy to forget when the code is modified later.

d2r commented on storm-core/src/jvm/org/apache/storm/container/cgroup/CgroupManager.java:
```diff
@@ -211,4 +209,15 @@ public void releaseResourcesForWorker(String workerId) {
+            LOG.warn("cgroup {} doesn't exist!", workerGroup);
+            return null;
+        }
+        Set<Long> ret = new HashSet<>();
```
`ret` isn't used here, but we could return an empty set instead of null.

d2r commented on storm-core/src/jvm/org/apache/storm/daemon/supervisor/AdvancedFSOps.java:
```diff
@@ -0,0 +1,202 @@
+     * @param conf the configuration of the process
+     * @return the appropriate instance of the class for this config and environment.
+     */
+    public static AdvancedFSOps mk(Map<String, Object> conf) {
```
`mk` is a little terse for non-clojure. Can we change it to `make`?

d2r commented on storm-core/src/jvm/org/apache/storm/daemon/supervisor/AdvancedFSOps.java:
```diff
@@ -0,0 +1,202 @@
+        
+        public AdvancedRunAsUserFSOps(Map<String, Object> conf) {
+            if (Utils.isOnWindows()) {
+                throw new RuntimeException("ERROR: Windows doesn't support running workers as different users yet");
```
UnsupportedOperationException

d2r commented on storm-core/src/jvm/org/apache/storm/daemon/supervisor/AdvancedFSOps.java:
```diff
@@ -0,0 +1,202 @@
+    
+    /**
+     * Setup permissions properly for an internal blob store path
+     * @param path the path to the permissions
```
path to blobstor?

d2r commented on storm-core/src/jvm/org/apache/storm/daemon/supervisor/Container.java:
```diff
@@ -0,0 +1,437 @@
+    public LSWorkerHeartbeat readHeartbeat() throws IOException {
+        LocalState localState = ConfigUtils.workerState(_conf, _workerId);
+        LSWorkerHeartbeat hb = localState.getWorkerHeartBeat();
+        LOG.warn("{}: Reading heartbeat {}", _workerId, hb);
```
debug

d2r commented on storm-core/src/jvm/org/apache/storm/daemon/supervisor/Container.java:
```diff
@@ -0,0 +1,437 @@
+    
+    private boolean isWindowsProcessAlive(long pid, String user) throws IOException {
+        boolean ret = false;
+        ProcessBuilder pb = new ProcessBuilder("tasklist", "/nh", "/fi", "pid eq"+pid);
```
Missing a space after `eq` here?

d2r commented on storm-core/src/jvm/org/apache/storm/daemon/supervisor/Container.java:
```diff
@@ -0,0 +1,437 @@
+        boolean allDead = true;
+        for (Long pid: pids) {
+            if (!isProcessAlive(pid, user)) {
+                LOG.warn("{}: PID {} is dead", _workerId, pid);
```
debug?

d2r commented on storm-core/src/jvm/org/apache/storm/daemon/supervisor/Container.java:
```diff
@@ -0,0 +1,437 @@
+     */
+    protected void setup() throws IOException {
+        if (_port <= 0) {
+            throw new IllegalStateException("Cannot setup a container recovered with just a worker id");
```
Could we clarify this a bit? I wasn't exactly sure what this meant with this wording.

d2r commented on storm-core/src/jvm/org/apache/storm/daemon/supervisor/Container.java:
```diff
@@ -0,0 +1,437 @@
+        final String topologyId = _assignment.get_topology_id();
+        if (!SupervisorUtils.doRequiredTopoFilesExist(_conf, topologyId)) {
+            LOG.info("Missing topology storm code, so can't launch  worker with assignment {} for this supervisor {} on port {} with id {}", _assignment,
+                    _supervisorId, _port, _workerId);
```
LOG.error, and maybe just say missing required files? We don't know specifically what was missing.

d2r commented on storm-core/src/jvm/org/apache/storm/daemon/supervisor/Container.java:
```diff
@@ -0,0 +1,437 @@
+    @SuppressWarnings("unchecked")
+    protected void writeLogMetadata(Map<String, Object> topologyConf, String user, String topologyId) throws IOException {
+        if (_port <= 0) {
+            throw new IllegalStateException("Cannot setup a container recovered with just a worker id");
```
cut & paste?

d2r commented on storm-core/src/jvm/org/apache/storm/daemon/supervisor/Container.java:
```diff
@@ -0,0 +1,437 @@
+            List<String> groups = (List<String>) topologyConf.get(Config.LOGS_GROUPS);
+            for (String group : groups){
+                logsGroups.add(group);
+            }
```
We could use addAll instead of looping, as we do below, here and elsewhere.

d2r commented on storm-core/src/jvm/org/apache/storm/daemon/supervisor/Container.java:
```diff
@@ -0,0 +1,437 @@
+        data.put("worker-id", _workerId);
+
+        Set<String> logsGroups = new HashSet<>();
+        //for supervisor-test
```
Why is this needed only for the test?

revans2 replied:
Copy and paste really don't know

d2r commented on storm-core/src/jvm/org/apache/storm/daemon/supervisor/Container.java:
```diff
@@ -0,0 +1,437 @@
+     */
+    protected void createArtifactsLink(String topologyId) throws IOException {
+        if (_port <= 0) {
+            throw new IllegalStateException("Cannot setup a container recovered with just a worker id");
```
copy/paste?

d2r commented on storm-core/src/jvm/org/apache/storm/daemon/supervisor/Container.java:
```diff
@@ -0,0 +1,437 @@
+     * @throws IOException on any error
+     */
+    public void cleanUpForRestart() throws IOException {
+        // and another API to cleanup with everything is dead
```
TODO?

d2r commented on storm-core/src/jvm/org/apache/storm/daemon/supervisor/ContainerLauncher.java:
```diff
@@ -0,0 +1,99 @@
+     * @return the proper container launcher
+     * @throws IOException on any error
+     */
+    public static ContainerLauncher mk(Map<String, Object> conf, String supervisorId, IContext sharedContext) throws IOException {
```
Rename `make`?

d2r commented on storm-core/src/jvm/org/apache/storm/daemon/supervisor/ContainerLauncher.java:
```diff
@@ -0,0 +1,99 @@
+        
+        ResourceIsolationInterface resourceIsolationManager = null;
+        if (Utils.getBoolean(conf.get(Config.STORM_RESOURCE_ISOLATION_PLUGIN_ENABLE), false)) {
+            resourceIsolationManager = Utils.newInstance((String) conf.get(Config.STORM_RESOURCE_ISOLATION_PLUGIN));
```
It seems `Config.STORM_RESOURCE_ISOLATION_PLUGIN` should have been declared as a String, but is an Object.  There's no harm, but it may be worth fixing it now.

revans2 replied:
done

d2r commented on storm-core/src/jvm/org/apache/storm/daemon/supervisor/ContainerLauncher.java:
```diff
@@ -0,0 +1,99 @@
+     * @return a Container that can be killed.
+     * @throws IOException on any error
+     */
+    public abstract Killable recoverContainer(String workerId) throws IOException;
```
I like the idea of using an interface specific to the scenario. Only comment is we should update the javadoc to say it returns a Killable instead of a Container.

abellina commented on storm-core/src/jvm/org/apache/storm/daemon/supervisor/AdvancedFSOps.java:
```diff
@@ -0,0 +1,300 @@
+     * @throws IOException on any error.
+     */
+    public void deleteIfExists(File path, String user, String logPrefix) throws IOException {
+        LOG.debug("Deleting path {}", path);
```
this method doesn't use the user or logPrefix. It also does the same thing as the overload below.

revans2 replied:
It does in the runasuserfsops.  There are two inner sub classes.

kishorvpatil replied:
Again, can this be info level log entry?

abellina commented on storm-core/src/jvm/org/apache/storm/daemon/supervisor/BasicContainer.java:
```diff
@@ -0,0 +1,569 @@
+     * @return the command to run for profiling.
+     */
+    private List<String> mkProfileCommand(ProfileAction action, boolean stop, String workerPid, String targetDir) {
+        if (action == ProfileAction.JMAP_DUMP) {
```
could we make this a switch statement with an if for JPROFILE_STOP?

revans2 replied:
I copied and pasted this, but I agree a switch would be cleaner.

revans2 commented:
@d2r @abellina @harshach @knusbaum 

I believe that I am done and that the code is ready to merge.  I have addressed all of the review comments to date.  I have finished with updating the unit tests.  I am going to start doing more extensive testing of CGROUP and run as user support, but because we are not too close to a 2.x release I think we can address any issues I find in follow up JIRA.

Please take another look and let me know.

kishorvpatil commented on storm-core/src/jvm/org/apache/storm/daemon/supervisor/AdvancedFSOps.java:
```diff
@@ -0,0 +1,313 @@
+        @Override
+        public void deleteIfExists(File path, String user, String logPrefix) throws IOException {
+            String absolutePath = path.getAbsolutePath();
+            LOG.debug("Deleting path {}", absolutePath);
```
This can probably be info level.

abellina commented on storm-core/src/jvm/org/apache/storm/daemon/supervisor/AdvancedFSOps.java:
```diff
@@ -0,0 +1,313 @@
+        return FileUtils.readFileToString(location, "UTF-8");
+    }
+    
+    public byte[] slurp(File location) throws IOException {
```
nit, missing a function header.

abellina commented on storm-core/src/jvm/org/apache/storm/daemon/supervisor/BasicContainer.java:
```diff
@@ -0,0 +1,629 @@
+            log4jConfigurationDir = stormHome + Utils.FILE_PATH_SEPARATOR + "log4j2";
+        }
+        
+
```
nit, blank line

abellina commented on storm-core/src/jvm/org/apache/storm/daemon/supervisor/BasicContainer.java:
```diff
@@ -0,0 +1,629 @@
+     * @param b something else
+     * @return a or b the first one that is not null
+     */
+    private <V> V OR(V a, V b) {
```
Should this be in util?

revans2 replied:
If others need it we can move it then.

abellina commented on storm-core/src/jvm/org/apache/storm/daemon/supervisor/BasicContainer.java:
```diff
@@ -0,0 +1,629 @@
+        }
+        topEnvironment.put("LD_LIBRARY_PATH", jlp);
+
+        // {"cpu" cpu "memory" (+ mem-onheap mem-offheap (int (Math/ceil (conf
```
remove?

revans2 replied:
This was here before and I think it is worth saving.

revans2 replied:
Never mind I'll remove it.

abellina commented on storm-core/src/jvm/org/apache/storm/daemon/supervisor/Container.java:
```diff
@@ -0,0 +1,493 @@
+    }
+    
+    private boolean isWindowsProcessAlive(long pid, String user) throws IOException {
+        boolean ret = false;
```
should this check the pids user, just like we check in the posix case?

revans2 replied:
I would like to but I don't really have access to a windows box so I didn't feel comfortable parsing the output of the windows function without testing it.

srdo replied:
On my Windows 10 install, tasklist only outputs these fields: "Image Name, PID, Session Name, Session#, Mem Usage", but there's a filter to get only processes started by a user. For example: 
`C:\Users\Stig>tasklist /nh /fi "pid eq 10300" /fi "username eq Stig"`
outputs
`bash.exe                     10300 Console                   21      1.484 K`

srdo replied:
Sorry, was a bit too fast there. tasklist also outputs user name in domain\username format if passed the /v flag.
`C:\Users\Stig>tasklist /nh /fi "pid eq 10300" /v`
outputs
`bash.exe                     10300 Console                   21        888 K Unknown         STIG-LAPTOP\Stig                                        0:00:00 N/A` 

The columns are "Image Name, PID, Session Name, Session#, Mem Usage, Status, User Name, CPU Time, Window Title"

revans2 replied:
Parsing this out without a working windows box is more then I am comfortable with.  If someone else wants to write it and make a pull request to this branch I am happy to accept it.

srdo replied:
I took at stab at it https://github.com/revans2/incubator-storm/pull/6

abellina commented on storm-core/src/jvm/org/apache/storm/daemon/supervisor/ReadClusterState.java:
```diff
@@ -0,0 +1,318 @@
+                    if (!portLA.containsKey(port)) {
+                        portLA.put(port, la);
+                    } else {
+                        throw new RuntimeException("Should not have multiple topologys assigned to one port");
```
nit, topologies

abellina commented on storm-core/src/jvm/org/apache/storm/daemon/supervisor/Slot.java:
```diff
@@ -0,0 +1,769 @@
+        }
+    }
+    
+    static boolean equivilant(LocalAssignment a, LocalAssignment b) {
```
nit, equivalent?

abellina commented on storm-core/src/jvm/org/apache/storm/daemon/supervisor/Slot.java:
```diff
@@ -0,0 +1,769 @@
+                    Map<Long, LocalAssignment> orig = null;
+                    do {
+                        Long lport = new Long(staticState.port);
+                        orig = cachedCurrentAssignmants.get();
```
cachedCurrentAssignmants -> cachedCurrentAssignments

abellina commented on storm-core/src/jvm/org/apache/storm/daemon/supervisor/Slot.java:
```diff
@@ -0,0 +1,769 @@
+                port,
+                iSupervisor,
+                localState);
+        this.newAssignment.set(dynamicState.newAssignment);
```
should call setNewAssignment?

revans2 replied:
We could, but that is meant to be a user facing API, and so I don't see a lot of value in trying to maintain the abstraction here.

abellina commented on storm-core/src/jvm/org/apache/storm/daemon/supervisor/Slot.java:
```diff
@@ -0,0 +1,769 @@
+                    copy.removeAll(removed);
+                } while (!profiling.compareAndSet(orig, copy));
+                dynamicState = nextState;
+                newAssignment.compareAndSet(localNewAssignment, dynamicState.newAssignment);
```
why is this compareAndSet? Also, in what case shouldn't newAssignment's current value not be == localNewAssignment? Maybe what I am missing is a comment.

revans2 replied:
This is where all of the threading is hidden away from the state machine.  There are only a few APIs that interact with slot asynchronously.  setNewAssignment is one of them, and profiling requests is the other.  To hide this from the state machine we take a snapshot of the new assignment before running the state machine, and then before we overwrite the result we check to see if it didn't change in the background.

abellina replied:
Ok. Thanks! Took me a while to go through Slot and the state machine. The unit tests were very helpful on that.

To summarize how I understand this:
1. we get a copy of new assignment
2. mutate localstate assignments
3. mutate supervisor in-memory assingments
4. deal with profiling cleanup
5. persist the new assignment in slot.
6. If new assignment persisted successfully, then we are good. The equivalency checks will pass, and the state machine should progress or stay at RUNNING. Else, the equivalency checks will move us to correct state. Also this loop fixes up localstate and in-memory assignments continuously.

srdo replied:
I don't really understand this assignment. The localNewAssignment set in L726 is in L731 put into a DynamicState. The resulting state from doing a step in the state machine is then used to overwrite dynamicState in L759. I can't find anywhere dynamicState.newAssignment is changed in the state machine, so I don't understand how localNewAssignment and dynamicState.newAssignment can be different. Where/when does nextState have its newAssignment changed?

abellina commented on storm-core/src/jvm/org/apache/storm/localizer/LocalDownloadedResource.java:
```diff
@@ -0,0 +1,107 @@
+        
+        @Override
+        public boolean cancel(boolean mayInterruptIfRunning) {
+            //cancel not currently supported
```
currently or never? (given that this is the NoCancel future)

revans2 replied:
Cancel is difficult to support, but not impossible.  The only value it offers right now would be to possibly stop downloading a jar/blob early if an assignment changed in the middle of the download.  This is rare enough that I didn't feel it was worth the extra effort to do it at this time.

abellina replied:
OK. Thanks

revans2 commented:
@abellina I addressed all of your comments.  I also have manually run all of the unit tests and they are passing for me, despite the travis issues.

HeartSaVioR commented on storm-core/src/jvm/org/apache/storm/daemon/supervisor/AdvancedFSOps.java:
```diff
@@ -0,0 +1,319 @@
+     * @throws IOException on any error.
+     */
+    public void deleteIfExists(File path, String user, String logPrefix) throws IOException {
+        LOG.info("Deleting path {}", path);
```
We could just call deleteIfExists(path) since it's same.

revans2 replied:
Great catch.

HeartSaVioR commented on storm-core/src/jvm/org/apache/storm/daemon/supervisor/BasicContainer.java:
```diff
@@ -0,0 +1,628 @@
+            createNewWorkerId();
+        }
+
+        String stormHome = System.getProperty("storm.home");
```
Minor: `System.getProperty("storm.home")` might be better to place to Utils or so since there're some other places which uses this. I'm also OK to keep this as it is and address this from other issue.

HeartSaVioR commented:
I'm still reviewing this. Nice work to refactor to introduce greater readability. Will leave comment here once I'm done with first pass.

revans2 commented:
@HeartSaVioR take your time.  I want to be sure that I have plenty of eyeballs looking at this.  We are doing this mostly because we started to run into a lot of race conditions in the supervisor.  We would try to fix one, and another would be exposed.  This makes it so that interacting with the container is only ever done on a single thread so we can reason about the state of the container and avoid those races.

srdo commented on storm-core/src/jvm/org/apache/storm/daemon/supervisor/AdvancedFSOps.java:
```diff
@@ -0,0 +1,314 @@
+        }
+        
+        @Override
+        public void moveDriectoryPreferAtomic(File fromDir, File toDir) throws IOException {
```
This function name is misspelled

srdo commented on storm-core/src/jvm/org/apache/storm/daemon/supervisor/AdvancedFSOps.java:
```diff
@@ -0,0 +1,314 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class AdvancedFSOps {
```
If this class is supposed to be subclassed before use, why not make it abstract?

revans2 replied:
AdvancedFSOps is the default only if you are on windows or if you are doing run as user is a subclass used.

HeartSaVioR commented on storm-core/src/jvm/org/apache/storm/daemon/supervisor/LocalContainer.java:
```diff
@@ -0,0 +1,85 @@
+    
+    @Override
+    public void launch() throws IOException {
+        IFn mkWorker = Clojure.var("org.apache.storm.daemon.worker", "mk-worker");
```
Let's leave a comment that it should be modified when worker is being ported.

srdo commented on storm-core/src/jvm/org/apache/storm/daemon/supervisor/BasicContainer.java:
```diff
@@ -0,0 +1,644 @@
+
+    /**
+     * Create a new worker ID for this process and store in in this object and
+     * in the local state.  Never call this if a worker is currently up and running.
```
Could this check if it is called at the wrong time? For example by throwing an IllegalStateException if _workerId is not null?

revans2 replied:
done

srdo commented on storm-core/src/jvm/org/apache/storm/daemon/supervisor/BasicContainer.java:
```diff
@@ -0,0 +1,644 @@
+     * @param conf the config for the supervisor.
+     * @return the java.library.path/LD_LIBRARY_PATH to use so native libraries load correctly.
+     */
+    protected String jlp(String stormRoot, Map<String, Object> conf) {
```
Nitpick: Consider not abbreviating this. It's obvious what it is here, but probably less so when called from elsewhere in the code.

srdo commented on storm-core/src/jvm/org/apache/storm/daemon/supervisor/BasicContainer.java:
```diff
@@ -0,0 +1,644 @@
+        
+        List<String> commonParams = new ArrayList<>();
+        commonParams.add("-Dlogfile.name=worker.log");
+        commonParams.add("-Dstorm.home=" + stormHome);
```
Could we put some of these properties in constants somewhere?

revans2 commented:
I think I have addressed most of the issues so far.  I have been running some manual tests and have run a cluster with run as user and cgroup enforcement enabled.  I plan on doing some more tests, but I think the current pull request should be ready for anyone who whats to take a look at it/test it to do so.

It should be a drop in replacement for the current supervisor, and I have even tested it doing a rolling upgrade/downgrade.

HeartSaVioR commented on storm-core/src/jvm/org/apache/storm/utils/ConfigUtils.java:
```diff
@@ -512,6 +463,10 @@ public static String workerTmpRoot(Map conf, String id) {
     }
+    
+    public static String workerPidPath(Map<String, Object> conf, String id, long pid) {
+        return (workerPidsRoot(conf, id) + FILE_SEPARATOR + pid);
```
Nit: might be better to call workerPidPath(conf, id, String.valueOf(pid)) or vice versa.

HeartSaVioR commented on storm-core/src/jvm/org/apache/storm/utils/ConfigUtils.java:
```diff
@@ -353,25 +350,21 @@ public LocalState nimbusTopoHistoryStateImpl(Map conf) throws IOException {
-    }
-
-    public StormTopology readSupervisorTopologyImpl(Map conf, String stormId) throws IOException {
+    public static StormTopology readSupervisorTopology(Map conf, String stormId, AdvancedFSOps ops) throws IOException {
```
Don't we need to consider mock like `readSupervisorStormConf()`? If we don't need to consider, let's remove above comment to make consistent.

revans2 replied:
I put in the AdvancedFSOps to handle the mocking, but now I am not so sure on that.  It is too low level for a lot of what we want to do, so I think I will put it back in, and just do the mocking at the higher level.

d2r replied:
Agreed. The comment describes a pattern that this change removes.

HeartSaVioR commented on storm-core/test/jvm/org/apache/storm/daemon/supervisor/BasicContainerTest.java:
```diff
@@ -0,0 +1,459 @@
+        assertEquals(new File(topoRoot), cmd.pwd);
+        
+        //JPROFILE START
+        req.set_action(ProfileAction.JPROFILE_STOP);
```
JPROFILE_STOP -> JPROFILE_START?

HeartSaVioR replied:
Never mind. I found that mkProfileCommand() only handles JPROFILE_STOP and determine whether start or stop by flag.

revans2 replied:
Ya it was really confusing until I asked @kishorvpatil about it and he explained it all to me.

HeartSaVioR commented on storm-core/src/jvm/org/apache/storm/utils/Utils.java:
```diff
@@ -1990,31 +1995,26 @@ protected void forceDeleteImpl(String path) throws IOException {
      */
-    public static void createSymlink(String dir, String targetDir,
-            String filename, String targetFilename) throws IOException {
+    public static void forceCreateSymlink(String dir, String targetDir,
```
The sequence of parameters have been swapped: targetFilename and filename.
Since parameters of parent directory are link and target, it would be natural to keep that sequence to filename, too.

revans2 replied:
Deleting it.  I added it then made a better version in fsops, so I am removing it.

HeartSaVioR commented:
I merged this in my local, and do some tests, and see:

1. integration test fails from my dev. machine

- I'm using OSX 10.11, Java 1.8.0_66
- I got wrong IP address from integration test. I'm using IP sharing router but assigned IP is even not same as external IP.
-- This is similar to what I consistently observed from my shared office. 
-- In my office even build for current master is failing. (I'm testing this pull request in home now.)
-- Strange thing is that normal unit tests are not failing for this build. In my office it's also failing.

2. There's a bug for creating symbolic link for dependencies, and also createSymbolicLink.  I'll leave a comment to diff.

HeartSaVioR commented on storm-core/src/jvm/org/apache/storm/daemon/supervisor/SupervisorUtils.java:
```diff
@@ -159,7 +162,7 @@ public static void addBlobReferences(Localizer localizer, String stormId, Map co
+    public static Set<String> readDownLoadedStormIds(Map<String, Object> conf) throws IOException {
         Set<String> stormIds = new HashSet<>();
         String path = ConfigUtils.supervisorStormDistRoot(conf);
         Collection<String> rets = Utils.readDirContents(path);
```
This is not a bug from the patch, but in OSX .DS_Store file can be created, so we need to filter out `files` in rets and keep only `directories`.

d2r replied:
I agree that it could be improved. Let's create a separate jira issue to track it. I think it is out-of-scope here.

HeartSaVioR commented:
Found other issue: 
Constructor of BasicContainer (in fact constructor of Container) throws FNF when topology dist files are deleted. So even though Slot is creating with recover mode, Supervisor is killed instead of throwing ContainerRecoveryException. Supervisor will consistently be killed unless we remove the assignment for that - killing topology.

This is a weird case but can be happening, and I guess that's why ContainerRecoveryException exists.

HeartSaVioR commented on storm-core/src/jvm/org/apache/storm/localizer/AsyncLocalizer.java:
```diff
@@ -0,0 +1,420 @@
+                    String blobSymlinkTargetName = new File(localizedResource.getCurrentSymlinkPath()).getName();
+
+                    String symlinkName = null;
+                    Map<String, Object> blobInfo = blobstoreMap.get(keyName);
```
If we use --jars or --dependencies and don't set blobstore map, blobstoreMap could be null.

Below code resolves NPE issue and make it works, but it might be better to clarify that it is better to include dependencies to blobstore map or keep this as it is.

```
if (blobstoreMap != null) {
                        Map<String, Object> blobInfo = blobstoreMap.get(keyName);
                        if (blobInfo != null && blobInfo.containsKey("localname")) {
                            symlinkName = (String) blobInfo.get("localname");
                        } else {
                            symlinkName = keyName;
                        }
                    } else {
                        // all things are from dependencies
                        symlinkName = keyName;
                    }
```

HeartSaVioR replied:
@revans2 It's broken again from latest version of the fix. (regression)

HeartSaVioR commented:
With my patch (symlink issue and NPE issue) I can see workers launched and killed by Supervisor V2. (remote)

Tested:
- kill worker process with -9
- rebalance with different worker count
- jstack dump, heap dump, restart worker

HeartSaVioR commented:
I found another issue: 
When I rebalance 3 workers into 1 worker, all workers are killed first (expected) and AsyncLocalizer clear out topology codes since all workers are killed. 
But Supervisor doesn't download topology code again while starting new worker, so it goes wrong and worker and/or supervisor are killed.

This seems to be a kind of race condition (Slot and AsyncLocalizer) and I saw two scenarios: 

1. Worker can be launched but topology directory is removed after launching so worker is crashed. Slot tries to relaunch worker but throws IllegalStateException because topology directory is gone and supervisor also be killed.

2. Supervisor is killed even before launching worker.

After this, Supervisor will consistently be killed unless clearing out supervisor directory as same as above comment.

d2r commented on storm-core/src/jvm/org/apache/storm/daemon/supervisor/Container.java:
```diff
@@ -0,0 +1,484 @@
+    
+    private boolean isWindowsProcessAlive(long pid, String user) throws IOException {
+        boolean ret = false;
+        ProcessBuilder pb = new ProcessBuilder("tasklist", "/nh", "/fi", "pid eq" + pid);
```
OK, I was not clear in my earlier comment. I didn't mean we needed spaces around the `+` but that we need one within the quoted string after the `eq`: `"pid eq " + pid`

revans2 replied:
great catch

d2r commented on storm-core/src/jvm/org/apache/storm/daemon/supervisor/ReadClusterState.java:
```diff
@@ -0,0 +1,318 @@
+                    }
+                    if (Time.isSimulating()) {
+                        Time.advanceTime(1000);
+                        Thread.sleep(100);
```
Should we advance time by the same amount as the sleep time?

revans2 replied:
We are simulating, so in general we don't actually wait as long in simulated time, to speed up the tests.

d2r replied:
OK good.

d2r commented on storm-core/src/jvm/org/apache/storm/daemon/supervisor/ReadClusterState.java:
```diff
@@ -0,0 +1,318 @@
+        this.superConf = superConf;
+        this.stormClusterState = stormClusterState;
+        this.syncSupEventManager = syncSupEventManager;
+        this.assignmentVersions = new AtomicReference<Map<String, VersionedData<Assignment>>>(new HashMap<String, VersionedData<Assignment>>());
```
very minor: This line is long and can be shorted by using <>  for the AtomicReference and for the HashMap.

revans2 replied:
I tried that, and for some reason java doesn't like it, because I am initializing it with a HashMap, that is a Map, but not exactly a Map. I wasn't totally sure why java didn't like it.

d2r replied:
OK.

srdo replied:
Changing it to `this.assignmentVersions = new AtomicReference<>(new HashMap<>());` compiles fine for me on Java 8.

revans2 replied:
Happy to be wrong here, it is a lot smaller.

d2r commented on storm-core/src/jvm/org/apache/storm/daemon/supervisor/ReadClusterState.java:
```diff
@@ -0,0 +1,318 @@
+            HashSet<Integer> allPorts = new HashSet<>(assignedPorts);
+            allPorts.addAll(slots.keySet());
+            
+            Map<Integer, Set<TopoProfileAction>> filtered = new HashMap<>();
```
minor: We define keys as Integer, but we are using Long for ports.

revans2 replied:
Ya that comes from thrift, and I really don't know why it is a Long there.

d2r commented on storm-core/src/jvm/org/apache/storm/daemon/supervisor/ReadClusterState.java:
```diff
@@ -0,0 +1,318 @@
+            Map<Integer, LocalAssignment> allAssignments =
+                    readAssignments(assignmentsSnapshot, assignmentId, readRetry);
+            if (allAssignments == null) {
+                //Something odd happened try again later
```
Is it worth adding a debug-level log here?

revans2 replied:
There already is one inside readAssignments.

d2r replied:
OK

d2r commented on storm-core/src/jvm/org/apache/storm/daemon/supervisor/ReadClusterState.java:
```diff
@@ -0,0 +1,318 @@
+            LOG.debug("All assignment: {}", allAssignments);
+            LOG.debug("Topology Ids -> Profiler Actions {}", topoIdToProfilerActions);
+            for (Integer port: allAssignments.keySet()) {
+                if (iSuper.confirmAssigned(port)) {
```
minor: We may want to use `this.` as we do above for private vars just to be consistent in the file. Not a big deal though.

revans2 replied:
We only do it in the constructor when the field names are the same.

d2r commented on storm-core/src/jvm/org/apache/storm/daemon/supervisor/ReadClusterState.java:
```diff
@@ -0,0 +1,318 @@
+                    if (!portLA.containsKey(port)) {
+                        portLA.put(port, la);
+                    } else {
+                        throw new RuntimeException("Should not have multiple topologies assigned to one port");
```
Would it be useful to include the topologies and the port in case we want to debug this?

d2r commented on storm-core/src/jvm/org/apache/storm/daemon/supervisor/ReadClusterState.java:
```diff
@@ -0,0 +1,318 @@
+            } else {
+                retries.addAndGet(1);
+            }
+            LOG.warn("{} : retrying {} of 3", e.getMessage(), retries.get());
```
minor, but if retries happens to get changed in another thread, we would be logging the incorrect retry count here.

revans2 replied:
This is code that came from clojure.  It might be worth rewriting the entire function to not use the AtomicInteger, but I thought it would be best to leave the code alone, and if we wanted to clean it up later we could.

d2r commented on storm-core/src/jvm/org/apache/storm/daemon/supervisor/ReadClusterState.java:
```diff
@@ -0,0 +1,318 @@
+                slot.setNewAssignment(null);
+            } catch (Exception e) {
+                LOG.error("Error trying to shutdown workers in {}", slot, e);
+            }
```
Are there Exceptions here we need to catch here?

revans2 replied:
:) good catch

d2r commented on storm-core/src/jvm/org/apache/storm/daemon/supervisor/RunAsUserContainer.java:
```diff
@@ -0,0 +1,107 @@
+            LocalState state, ResourceIsolationInterface resourceIsolationManager, boolean recover) throws IOException {
+        super(port, assignment, conf, supervisorId, state, resourceIsolationManager, recover);
+        if (Utils.isOnWindows()) {
+            throw new RuntimeException("ERROR: Windows doesn't support running workers as different users yet");
```
UnsupportedOperationException

d2r commented on storm-core/src/jvm/org/apache/storm/daemon/supervisor/RunAsUserContainer.java:
```diff
@@ -0,0 +1,107 @@
+            ResourceIsolationInterface resourceIsolationManager) throws IOException {
+        super(workerId, conf, supervisorId, resourceIsolationManager);
+        if (Utils.isOnWindows()) {
+            throw new RuntimeException("ERROR: Windows doesn't support running workers as different users yet");
```
UnsupportedOperationException, and in fact we could add a Utils#unsupportedOnWindows method and just call it as we would a validator.

d2r commented on storm-core/src/jvm/org/apache/storm/daemon/supervisor/RunAsUserContainer.java:
```diff
@@ -0,0 +1,107 @@
+        //With run as user we do the isolation before switching users, so don't modify the original
+        // command line
+        return command;
+    }   
```
@revans2 , Let's move updateCommandForIsolation into launchWorkerProcess, as we discussed. It will clean up the code.

d2r commented on storm-core/src/jvm/org/apache/storm/daemon/supervisor/RunAsUserContainerLauncher.java:
```diff
@@ -0,0 +1,71 @@
+            // We could not recover return null
+        }
+        return container;
+    }
```
Could just `return RunAsUserContainer(w, c, s, r);`

revans2 replied:
But then the ContainerRecoveryException will not be turned into a null.  We can change that everywhere, but for now it is an implementation detail that is hidden.

d2r replied:
Yes, I botched my comment:  I wanted to suggest recoverContainer(p,a,s) could just call recoverContainer(w), but the two have different return types. So we should leave it.

d2r commented on storm-core/src/jvm/org/apache/storm/daemon/supervisor/Slot.java:
```diff
@@ -0,0 +1,766 @@
+                state == MachineState.KILL_AND_RELAUNCH) {
+                sb.append(" msInState: ");
+                sb.append(Time.currentTimeMillis() - startTime);
+            }
```
Although we don't need to track start times for every state for functionality, it might be nice to keep track of start times if only to print them in log messages. This is a nice-to-have suggestion and not required now. But it might be easy to add.

revans2 replied:
Originally I had it so that we would start the counter in one state and then time things out in a different state, but I combined states enough that we should be able to do it all the time.

d2r commented on storm-core/src/jvm/org/apache/storm/daemon/supervisor/Slot.java:
```diff
@@ -0,0 +1,766 @@
+    static boolean equivalent(LocalAssignment a, LocalAssignment b) {
+        if (a == null && b == null) {
+            return true;
+        } if (a != null && b != null) {
```
It looks like this used to be an `else if` but can now be outdented and move to its own line.

d2r commented on storm-core/src/jvm/org/apache/storm/daemon/supervisor/Slot.java:
```diff
@@ -0,0 +1,766 @@
+                    boolean bHasResources = b.is_set_resources();
+                    if (!aHasResources && !bHasResources) {
+                        return true;
+                    } else if (aHasResources && bHasResources) {
```
Likewise this can be an `if` on its own line.

d2r commented on storm-core/src/jvm/org/apache/storm/daemon/supervisor/Slot.java:
```diff
@@ -0,0 +1,766 @@
+        LSWorkerHeartbeat hb = dynamicState.container.readHeartbeat();
+        if (hb == null) {
+            LOG.warn("SLOT {}: HB returned as null", staticState.port);
+            //This should never happen, but to be safe
```
As we discussed, this could happen in the case when we successfully recover a container while the worker it contains never initialized far enough to write its first heartbeat. This is still an OK state to be in, so the comment should change.

d2r commented on storm-core/src/jvm/org/apache/storm/daemon/supervisor/Slot.java:
```diff
@@ -0,0 +1,766 @@
+            LOG.warn("SLOT {}: Container {} failed to launch in {} ms.", staticState.port, dynamicState.container, staticState.firstHbTimeoutMs);
+            dynamicState.container.kill();
+            Time.sleep(staticState.killSleepMs);
+            return dynamicState.withState(MachineState.KILL_AND_RELAUNCH);
```
We should just call `killAndRelaunchContainer` here. In this transition, there should not be any profiler actions as it has not come up yet, so that extra setup is a NOOP.

d2r replied:
This satisfies the precondition for `handleKillAndRelaunch`.

d2r commented on storm-core/src/jvm/org/apache/storm/daemon/supervisor/Slot.java:
```diff
@@ -0,0 +1,766 @@
+        assert(dynamicState.pendingDownload != null);
+        assert(dynamicState.container == null);
+        
+        //Ignore changes to scheduling while downloading the topology code
```
Might want to add the reasoning: It is because properly cancelling the other thread's downloads while other workers might need the same files adds a ton of complexity for an uncommon case.

d2r commented on storm-core/src/jvm/org/apache/storm/localizer/AsyncLocalizer.java:
```diff
@@ -0,0 +1,420 @@
+ * This is a wrapper around the Localizer class that provides the desired
+ * async interface to Slot.
+ * TODO once we have replaced the original supervisor merge this with
+ * Localizer and optimize them
```
Yes, we want to do this. We should track this with a jira instead.

d2r commented on storm-core/src/jvm/org/apache/storm/localizer/AsyncLocalizer.java:
```diff
@@ -0,0 +1,420 @@
+        }
+
+        @Override
+        public Void get(long timeout, TimeUnit unit) {
```
My IDE is warning me that the TimeUnit param needs a `@NotNull` annotation or else the requirement gets overridden here, which is probably not intended.

revans2 replied:
The only `@NotNull` I could find that java could be using is a sun internal annotation.

revans2 commented:
@HeartSaVioR Yes it looks like I need to think through recovery and any races with the AsyncLocalizer a bit more.  I'll try to reproduce your error.

d2r commented on storm-core/src/jvm/org/apache/storm/localizer/AsyncLocalizer.java:
```diff
@@ -0,0 +1,420 @@
+                if (deleteAll) {
+                    LOG.info("Failed to download basic resources for topology-id {}", _topologyId);
+                    Utils.forceDelete(tmproot);
+                    Utils.forceDelete(stormroot);
```
We can catch and ignore any IOException from these forceDelete calls so that the original exception bubbles up.

d2r commented on storm-core/src/jvm/org/apache/storm/localizer/AsyncLocalizer.java:
```diff
@@ -0,0 +1,420 @@
+                    LOG.warn("{} already downloaded blobs, skipping", _topologyId);
+                    return null;
+                }
+            }
```
Let's make a helper method for this `if`, since we do the same thing for the DownloadBaseBlobsLocal.

revans2 replied:
I refactored this instead so the code is shared by a single class.

d2r commented on storm-core/src/jvm/org/apache/storm/localizer/AsyncLocalizer.java:
```diff
@@ -0,0 +1,420 @@
+                    if (codeOutStream != null)
+                        codeOutStream.close();
+                    if (confOutStream != null)
+                        codeOutStream.close();
```
* braces
* catch IOException

d2r commented on storm-core/src/jvm/org/apache/storm/localizer/AsyncLocalizer.java:
```diff
@@ -0,0 +1,420 @@
+                    Utils.extractDirFromJar(resourcesJar, ConfigUtils.RESOURCES_SUBDIR, stormroot);
+                } else if (url != null) {
+                    LOG.info("Copying resources at {} to {} ", url.toString(), targetDir);
+                    if (url.getProtocol() == "jar") {
```
`.equals` here?

d2r commented on storm-core/src/jvm/org/apache/storm/localizer/AsyncLocalizer.java:
```diff
@@ -0,0 +1,420 @@
+                if (deleteAll) {
+                    LOG.info("Failed to download basic resources for topology-id {}", _topologyId);
+                    Utils.forceDelete(tmproot);
+                    Utils.forceDelete(stormroot);
```
IOException

d2r commented on storm-core/src/jvm/org/apache/storm/localizer/AsyncLocalizer.java:
```diff
@@ -0,0 +1,420 @@
+
+    @Override
+    public synchronized void releaseSlotFor(String topologyId, int port) throws IOException {
+        LOG.warn("Releaseing slot for {} {}", topologyId, port);
```
"Releasing"

d2r commented:
I was able to review most of the code.  I skimmed the tests and didn't see anything strange. 

@revans2 knows that I will not be able to log on again for awhile.  I want to leave a note for others that, as my comments were either minor or were something we discussed, I don't want to hold up a merge due to my unresolved review comments should this change otherwise be ready to merge before I am back.

revans2 commented:
@HeartSaVioR I figured out what happened to make the supervisor crash.  In the transition from RUNNING to KILL to speed things up slot starts localizing resources for the new assignment.  In the transition from KILL to WAITING_FOR_BASIC_LOCALIZATION slot releases the resources for the container that was just killed.  But the reference counting is only for the topology id and port.  If the port and topology id are the same, then we get this.  I need a key that is specific to the assignment/port, not the topology/port.  I'll try to figure out what that should look like.

abellina commented on storm-core/src/jvm/org/apache/storm/daemon/supervisor/Slot.java:
```diff
@@ -0,0 +1,768 @@
+        
+        if (!equivalent(dynamicState.newAssignment, dynamicState.currentAssignment)) {
+            //We were rescheduled while waiting for the worker to come up
+            return Slot.killContainerForChangedAssignment(dynamicState, staticState);
```
Add log line to make it consistent with other calls to killContainerForChangedAssignment. Also no need for Slot.

LOG.warn("SLOT {}: Assignment Changed from {} to {}", staticState.port, dynamicState.currentAs    signment, dynamicState.newAssignment);

revans2 commented:
I just pushed in a much of new fixes and addressed all of the outstanding review comments.  I think it is good to go.  Please take another look/test it and let me know.

I have run tests with rebalancing, rolling upgrades, Recovery, deleting directories from distcache, and added directories to distcache.

revans2 commented:
Sorry, git just ate all of my more recent changes, and I have no idea what happened.  I'll try to see if I can recover them, but I might have to start over...

srdo commented:
Check the ref log, maybe you can recover https://git-scm.com/docs/git-reflog

revans2 commented:
@srdo Ya that is what I did.  Thanks for the advice.  Not sure what happened somehow when I upmerged all but 2 of my commits disappeared.

revans2 commented:
I recovered it and fixed some issues with integration tests/rat.  I think it should be good to go.

revans2 commented:
Looks like there were two failures in travis.  One is a rat issue with storm-submit-tools the other is an integration test that timed out.  I'll try to reproduce the issues and see if I can fix them.

revans2 commented:
I could not reproduce the one failure and the rat failure is fixed by STORM-2054 https://github.com/apache/storm/pull/1648

revans2 commented:
Not sure what has been happening with travis not being able to get to the apache maven repo all the time, but my build in travis passed https://travis-ci.org/revans2/incubator-storm/builds/157725399 despite the official build failing.

revans2 commented on storm-core/src/jvm/org/apache/storm/daemon/supervisor/Supervisor.java:
```diff
@@ -17,135 +17,231 @@
+    private final String assignmentId;
+    private final String hostName;
+    // used for reporting used ports when heartbeating
+    private final AtomicReference<Map<Long, LocalAssignment>> currAssignment;
```
currAssignment is never set.  Only the supervisor hb to nimbus uses this, and nimbus is OK with having used nodes show up as free, so it really only impacts the UI, showing slots as both free and assigned at the same time.  Will fix it.

revans2 replied:
Actually it is just on restart of the supervisor.  It recovers the running topology but doesn't update it.

revans2 replied:
Just fixed it

revans2 commented:
@knusbaum @srdo @harshach @HeartSaVioR @abellina I really would appreciate it if you could take another look at this patch.  I think everything is ready to go except a final squash before merging it.  The unit tests all pass, I have run some basic manual tests with CGroups and run as user enabled.  I have run a number of manual tests on my mac, including rebalancing, killing the workers, killing the supervisor, and going back and forth between old and new supervisor.  I have manually deleted and added files/directories in storm-dist on the supervisor.  I also ran the full set of the internal yahoo integration tests against it and everything has passed.

I really would like to start rolling this out internally at yahoo, but I want to get as many eyes looking at it to find issues first.  Better here then in production.  On the upside the bugs I have found have been rather easy to debug, and very reproducible because the race conditions have been eliminated when interacting with the Container.

abellina commented on storm-core/src/jvm/org/apache/storm/daemon/supervisor/Slot.java:
```diff
@@ -0,0 +1,780 @@
+                    this.pendingStopProfileActions);
+        }
+        
+        public DynamicState withPendingLocalization(Future<Void> pendingDownload) {
```
nit, could return: withPendingLocalization(this.pendingLocalization, pendingDownload);

abellina commented on storm-core/src/jvm/org/apache/storm/daemon/supervisor/Slot.java:
```diff
@@ -0,0 +1,780 @@
+        
+        @Override
+        public String toString() {
+            return "{ "+topoId + ": " + request + " }";
```
nit, spaces around first +.

abellina commented on storm-core/src/jvm/org/apache/storm/localizer/AsyncLocalizer.java:
```diff
@@ -0,0 +1,427 @@
+
+        List<String> rtn = new ArrayList<String>();
+        int size = jarPaths.size();
+        for (int i = 0; i < size; i++) {
```
how about: for (String path : jarPaths)

HeartSaVioR commented:
I've left a comment of regression issue, but other manual tests passed including failed things.
I'd be +1 once regression issue is resolved. Great works.

HeartSaVioR commented:
And I saw some places using lambda which can't be applied to 1.x. 
@revans2 Could you craft a pull request for 1.x branch (might also 1.0.x) too when we're sure it's good to merge?

harshach commented:
@HeartSaVioR do we want this for 1.x branch? Given that it's a java migration and pretty much a new feature that can get into 1.x-branch. I would rather keep 1.x-branch as it is.
@revans2 will go through it today.

revans2 commented:
@HeartSaVioR @harshach Once this goes in I would be happy to make a version for 1.x

In fact we are in the process of pulling it back for the version we use internally, and I was able to keep the old supervisor in place as a backup option.

As for the regression I will try and find the original code and see where I made a mistake.

HeartSaVioR commented:
@harshach There's already an another pull request for fixing race condition on old supervisor. Given that we can't set a due date for 2.0 (and maintenance date for 1.x) it would be better to have this as kind of bugfix.

@revans2 The code block in my comment fixes the issue. You seem to fix it, but it's gone now.

revans2 commented:
@HeartSaVioR I found the original comment and put it the change.  I also fixed some other review comments from @abellina and also fixed an annoying failure with drpc_auth_test.clj not being able to bind to a port.

HeartSaVioR commented:
@revans2 Thanks for the great work. Really appreciated. +1

revans2 commented:
The Travis failures are from hive dependencies not being downloaded.

revans2 commented:
@harshach did you get a chance to take a look yet?

srdo commented on storm-core/src/jvm/org/apache/storm/daemon/supervisor/Supervisor.java:
```diff
@@ -17,135 +17,231 @@
+        this.eventManager = new EventManagerImp(false);
+        this.readState = new ReadClusterState(this);
+        
+        Set<String> downloadedStormIds = SupervisorUtils.readDownLoadedStormIds(conf);
```
Nitpick: Download isn't two words

srdo commented on storm-core/src/jvm/org/apache/storm/daemon/supervisor/Supervisor.java:
```diff
@@ -163,15 +259,94 @@ public Integer call() throws Exception {
+        for (Killable k: containers) {
+            try {
+                k.forceKill();
+                while(!k.areAllProcessesDead()) {
```
Just in case a bug is introduced at some point, you may want to cap how long this is allowed to loop before it starts error logging

srdo commented on storm-core/src/jvm/org/apache/storm/localizer/AsyncLocalizer.java:
```diff
@@ -0,0 +1,422 @@
+
+        @Override
+        public Void call() throws Exception {
+            String stormroot = ConfigUtils.supervisorStormDistRoot(_conf, _topologyId);
```
You should probably consider throwing a try-catch Exception or even Throwable around these callables for logging, in case there are bugs. I think ExecutorService swallows exceptions by default.

revans2 replied:
The ExecutorService will store the Exception in the Future associated with making the call.  It will be thrown when .get is called on the Future.  We are doing that in Slot.  So in theory we should not miss anything, but yes you are right in practice things can be missed, and it is a huge pain when a bug is introduced and Exceptions are eaten. I'll look into it.

srdo replied:
You're right, it should be fine. I've mostly been using ExecutorService in a context where we didn't call .get (f. ex. scheduling), so I may be a little trigger happy :)

srdo commented on storm-core/src/jvm/org/apache/storm/daemon/supervisor/BasicContainer.java:
```diff
@@ -0,0 +1,644 @@
+     * and not a partial recovery the real conf is read.
+     * @param profileCmd the command to use when profiling (used for testing)
+     */
+    public BasicContainer(ContainerType type, Map<String, Object> conf, String supervisorId, int port,
```
Extremely nitpicky, but if the last 3 parameters are just for testing, consider making a package private overload of this and removing them from the public constructor

revans2 replied:
Yes, but the subclasses need access to them too for testing...  I guess they are all in the same package at this point so I'll see how it goes.

revans2 commented on storm-core/src/jvm/org/apache/storm/daemon/supervisor/ContainerLauncher.java:
```diff
@@ -0,0 +1,103 @@
+     * @return The container that can be used to manage the processes.
+     * @throws IOException on any error
+     */
+    public abstract Container recoverContainer(int port, LocalAssignment assignment, LocalState state) throws IOException;
```
Yes I have been meaning to switch it over to throw an exception instead of returning null.  Will do that

revans2 commented on storm-core/src/jvm/org/apache/storm/daemon/supervisor/BasicContainer.java:
```diff
@@ -0,0 +1,644 @@
+ */
+public class BasicContainer extends Container {
+    private static final Logger LOG = LoggerFactory.getLogger(BasicContainer.class);
+    private static final FilenameFilter jarFilter = new FilenameFilter() {
```
Agreed, This was a copy and paste, but you are right it would be cleaner.

revans2 commented on storm-core/src/jvm/org/apache/storm/daemon/supervisor/BasicContainer.java:
```diff
@@ -0,0 +1,644 @@
+                    }
+                }
+                if (wid == null) {
+                    throw new ContainerRecoveryException("Could not find worker id for " + port + " " + assignment);
```
The exception is actually OK, it will result in the slot being Empty which is logged already.

srdo replied:
Great :)

ptgoetz commented:
@revans2 I'm also in the process of reviewing this. I hope to finish by the end of the day. So far everything looks good to me.

revans2 commented on storm-core/src/jvm/org/apache/storm/daemon/supervisor/BasicContainer.java:
```diff
@@ -0,0 +1,644 @@
+            return runProfilingCommand(command, env, logPrefix, targetFile);
+        }
+        LOG.warn("PROFILING REQUEST NOT SUPPORTED {} IGNORED...", request);
+        return true;
```
This is a special case.  It essentially means that someone created a profiling request that we don't know what to do with.  This could be because of a rolling upgrade or something like that, but somehow nimbus did something we don't expect.

Returning false will cause the command to be retried, but at this point the supervisor has no idea what to do with the profiling request and the original code was ignoring these, so I decided to do the same.  If we want to change it we can, but I would rather do that in a follow on JIRA.

srdo replied:
Oh right, that makes sense. I wasn't looking at the place this is used, just went by the Javadoc for the overridden method.

revans2 commented on storm-core/src/jvm/org/apache/storm/daemon/supervisor/BasicContainer.java:
```diff
@@ -0,0 +1,644 @@
+        if (environment != null) {
+            topEnvironment.putAll(environment);
+        }
+        topEnvironment.put("LD_LIBRARY_PATH", jlp);
```
If I have `JNI.so` and it depends on `someLibrary.so` without `LD_LIBRARY_PATH` `JNI.so` will not be able to find  `someLibrary.so` if both were shipped in the jar together.

srdo replied:
Makes sense.

revans2 commented on storm-core/src/jvm/org/apache/storm/daemon/supervisor/ReadClusterState.java:
```diff
@@ -0,0 +1,328 @@
+public class ReadClusterState implements Runnable, AutoCloseable {
+    private static final Logger LOG = LoggerFactory.getLogger(ReadClusterState.class);
+    
+    private final Map<String, Object> superConf;
```
I was originally trying to follow the convention in the file itself.  This came from Supervisor.java that didn't do it and was split off from that as part of the review process.  I personally like the _ convention, but am OK with whatever.

srdo replied:
Alright, it doesn't really matter to me, just thought I'd mention it.

revans2 commented on storm-core/src/jvm/org/apache/storm/daemon/supervisor/ReadClusterState.java:
```diff
@@ -0,0 +1,328 @@
+    }
+    
+    protected Map<Integer, LocalAssignment> readAssignments(Map<String, VersionedData<Assignment>> assignmentsSnapshot,
+            String assignmentId, AtomicInteger retries) {
```
This came from the original clojure code, that was then translated into java, and I just left it.  The way it works right now is that there is a sleep in between the retries in the main thread.  I wanted to fix it, but felt leaving it would be simpler.  If you want me to I will change it.

srdo replied:
It wasn't so much that it uses a retries counter, I was just wondering why the method doesn't refer to the readRetry field directly. It's probably fine like this.

revans2 replied:
Oh that was when we had a SupervisorData object that was separate from the Supervisor.  I'll update it to not pass it in.

revans2 commented on storm-core/src/jvm/org/apache/storm/daemon/supervisor/Slot.java:
```diff
@@ -0,0 +1,776 @@
+     * @throws IOException on any error
+     */
+    static DynamicState prepareForNewAssignmentOnEmptySlot(DynamicState dynamicState, StaticState staticState) throws IOException {
+        assert(dynamicState.container == null);
```
The state itself is not necessarily empty.  It could be that we just killed all of the workers.  I'll rename the function to avoid confusion.

revans2 commented on storm-core/src/jvm/org/apache/storm/daemon/supervisor/Slot.java:
```diff
@@ -0,0 +1,776 @@
+    static DynamicState cleanupCurrentContainer(DynamicState dynamicState, StaticState staticState, MachineState nextState) throws Exception {
+        assert(dynamicState.container != null);
+        assert(dynamicState.currentAssignment != null);
+        
```
Yes, but that has side effects (it runs a process).  If you really want it I will do it.

srdo replied:
Does that matter if it's only evaluated when running tests?

revans2 replied:
That is the thing I really don't like putting anything into an assert with side effects, because it will behave differently between testing and production.  In this case it really should not matter, because the side effect is minimal, and it is just timing.

srdo replied:
Makes sense. Feel free to leave it out.

revans2 commented on storm-core/src/jvm/org/apache/storm/daemon/supervisor/Slot.java:
```diff
@@ -0,0 +1,776 @@
+    static DynamicState handleWaitingForBlobLocalization(DynamicState dynamicState, StaticState staticState) throws Exception {
+        assert(dynamicState.pendingLocalization != null);
+        assert(dynamicState.pendingDownload != null);
+        assert(dynamicState.container == null);
```
But the state is WaitingForBlobLocalization.  That is how this function is called.

srdo replied:
Right, sorry. I misunderstood the Javadoc. Should have paid more attention :)

revans2 commented on storm-core/src/jvm/org/apache/storm/daemon/supervisor/Slot.java:
```diff
@@ -0,0 +1,776 @@
+    static DynamicState handleWaitingForBasicLocalization(DynamicState dynamicState, StaticState staticState) throws Exception {
+        assert(dynamicState.pendingLocalization != null);
+        assert(dynamicState.pendingDownload != null);
+        assert(dynamicState.container == null);
```
Again the only way this is called is if the state is WaitingForBasicLocalization.

revans2 commented on storm-core/src/jvm/org/apache/storm/daemon/supervisor/Slot.java:
```diff
@@ -0,0 +1,776 @@
+                return dynamicState.withState(MachineState.WAITING_FOR_WORKER_START);
+            }
+            //Scheduling changed after we killed all of the processes
+            return prepareForNewAssignmentOnEmptySlot(cleanupCurrentContainer(dynamicState, staticState, null), staticState);
```
I guess that might clean things up to just go to empty directly.  I have tested this and would prefer to leave it, but if you feel strongly about it I will make the change.

srdo replied:
Feel free to leave it. It shouldn't really matter.

revans2 commented:
@srdo on the  `newAssignment.compareAndSet` you are 100% correct.  I originally had it in there to clear out the assignment when it is processed, but then realized that each time a new assignment is downloaded it would go back to what it was before and we needed to handle that. I'll clean it up.

revans2 commented:
@srdo Pushed a new version that hopefully addresses your review comments.

revans2 commented:
Not sure why travis has so many issues with the build and maven but here is my travis setup on the same branch with the build passing.

https://travis-ci.org/revans2/incubator-storm/builds/160266388

ptgoetz commented:
I have a few style-related nits, but I've refrained from pointing them out because the style(s) in the codebase are all over the place, and the style is inherited in some places. If we want to get nit picky over style, we should really have a style guide in place. That's a JIRA for another day...

Overall I think it looks great and I am +1 for merging. Thanks @revans2 for the great work, and everyone else for the thorough review.

In terms of back porting this to other branches, I'm okay with that. While the impetus behind this work was the Clojure to Java migration, I think the improvements made here warrant back-porting to earlier version branches. It also doesn't really affect user-facing APIs, so that's not a concern.

srdo commented:
@revans2 Thanks, looks great. +1 for merging.

revans2 commented:
@harshach I have +1 from enough people to merge this in and would really like to move forward with it, but if you are still looking at it I want to be sure you get a chance to finish.  Please let me know either way.

revans2 commented:
@harshach it has been 6 hours so I am going to check this in because I don't want to have to try and keep it up to date as the code base changes.  If you do find anything as time goes on I will be very happy to address your concerns in other pull requests.

revans2 commented:
I merged this into master and squashed the commits so I am closing the pull request, but will leave the JIRA open so we can look into pulling it back into 1.x as well.
